ASM/c/chests.c

```diff
@@ -41,87 +43,157 @@ void get_chest_override(z64_actor_t *actor) {
+    if (CHEST_SIZE_MATCH_CONTENTS && chest_type == SILVER_CHEST)
+    {
         chest_type = GOLD_CHEST;
     }
 
-    //write matrix
-    Mtx_t *mtx = write_matrix_stack_top(gfx);
-    gSPMatrix((*opa_ptr)++, mtx, G_MTX_MODELVIEW | G_MTX_LOAD | G_MTX_NOPUSH);
+    return chest_type;
+}
+
+void set_chest_texture(z64_gfx_t *gfx, uint8_t chest_type, Gfx **opa_ptr) {
+    //set texture type
+    void* frontTexture = (void*)BROWN_FRONT_TEXTURE;
+    void* baseTexture = (void*)BROWN_BASE_TEXTURE;
 
-    int dlist;
+    if (CHEST_SIZE_TEXTURE || CHEST_TEXTURE_MATCH_CONTENTS) {
+        if (chest_type == GILDED_CHEST) {
+            frontTexture = &GILDED_CHEST_FRONT_TEXTURE;
+            baseTexture = &GILDED_CHEST_BASE_TEXTURE;
+        }
+        else if (chest_type == SILVER_CHEST) {
+            frontTexture = &SILVER_CHEST_FRONT_TEXTURE;
+            baseTexture = &SILVER_CHEST_BASE_TEXTURE;
+        }
+        else if (chest_type == SKULL_CHEST_SMALL || chest_type == SKULL_CHEST_BIG) {
+            frontTexture = &SKULL_CHEST_FRONT_TEXTURE;
+            baseTexture = &SKULL_CHEST_BASE_TEXTURE;
+        }
+    }
+
+    //the brown chest's base and lid dlist has been modified to jump to
+    //segment 09 in order to dynamically set the chest front and base textures
+    gfx->poly_opa.d -= 4;
+    gDPSetTextureImage(gfx->poly_opa.d, G_IM_FMT_RGBA, G_IM_SIZ_16b, 1, frontTexture);
+    gSPEndDisplayList(gfx->poly_opa.d + 1);
+    gDPSetTextureImage(gfx->poly_opa.d + 2, G_IM_FMT_RGBA, G_IM_SIZ_16b, 1, baseTexture);
+    gSPEndDisplayList(gfx->poly_opa.d + 3);
+
+    //push custom dlist (that sets the texture) to segment 09
+    gSPSegment((*opa_ptr)++, 0x09, gfx->poly_opa.d);
+}
+
+void draw_chest_base(z64_game_t *game, z64_actor_t *actor, Gfx **opa_ptr)
+{
+    z64_gfx_t *gfx = game->common.gfx;
+    uint8_t chest_type = get_chest_type(actor);
+    gSPMatrix((*opa_ptr)++, write_matrix_stack_top(gfx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+    if (chest_type != GOLD_CHEST)
+    {
+        set_chest_texture(gfx, chest_type, opa_ptr);
+        gSPDisplayList((*opa_ptr)++, 0x060006F0);
+    }
+    else
+    {
+        gSPDisplayList((*opa_ptr)++, 0x06000AE8);
+    }
+}
 
-    if (part == CHEST_BASE) {
-        if (chest_type == GOLD_CHEST)
-            dlist = 0x06000AE8;
-        else
-            dlist = 0x060006F0;
+void draw_chest_lid(z64_game_t *game, z64_actor_t *actor, Gfx **opa_ptr)
+{
+    z64_gfx_t *gfx = game->common.gfx;
+    uint8_t chest_type = get_chest_type(actor);
+    gSPMatrix((*opa_ptr)++, write_matrix_stack_top(gfx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+    if (chest_type != GOLD_CHEST)
+    {
+        set_chest_texture(gfx, chest_type, opa_ptr);
+        gSPDisplayList((*opa_ptr)++, 0x060010C0);
+    }
+    else
+    {
+        gSPDisplayList((*opa_ptr)++, 0x06001678);
+    }
+}
 
+void draw_chest(z64_game_t *game, int32_t part, void *unk, void *unk2, z64_actor_t *actor, Gfx **opa_ptr)
```

Comment:
FYI refactored this function to make it almost identical to the original source code. Also split draw_chest_base and draw_chest_lid into separate functions so they can be called from forest hallway code separately.